plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEnginePublishOperation.java

```diff
@@ -0,0 +1,270 @@
+   * specified server.
+   */
+  public LocalAppEnginePublishOperation(LocalAppEngineServerBehaviour server, int kind, IModule[] modules,
+      int deltaKind) {
+    super("Publish to server", "Publish modules to App Engine Development Server");
+    this.server = server;
+    this.modules = modules;
+    this.kind = kind;
+    this.deltaKind = deltaKind;
+    IPath base = server.getRuntimeBaseDirectory();
+    helper = new PublishHelper(base.toFile());
+  }
+
+  @Override
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    // TODO: use more advanced key to store modules publish locations?
+    // Because a dependent
+    // java project (added as child modules and published as jar) cannot
+    // present in more than one
+    // parent modules.
+    // (Is this TODO still valid?)
+    List<IStatus> statusList = Lists.newArrayList();
+    IPath deployPath = server.getModuleDeployDirectory(modules[0]);
+    if (modules.length == 1) {
+      // root modules
+      publishDirectory(deployPath, statusList, monitor);
+    } else {
+      // TODO: Why and when does this code happen?
+      for (int i = 0; i < modules.length - 1; i++) {
+        IWebModule webModule = (IWebModule) modules[i].loadAdapter(IWebModule.class, monitor);
+        if (webModule == null) {
+          statusList.add(newErrorStatus("Not a Web module: " + modules[i].getName()));
+          return;
+        }
+        String uri = webModule.getURI(modules[i + 1]);
+        if (uri != null) {
+          deployPath = deployPath.append(uri);
+        } else {
+          // no uri is OK for removed modules
+          if (deltaKind != ServerBehaviourDelegate.REMOVED) {
+            statusList.add(newErrorStatus("Cannot get URI for module: " + modules[i + 1].getName()));
+            return;
+          }
+        }
+      }
+      // TODO: Disabled saving and loading module locations below (marked
+      // with XXX)
+
+      // modules given as parent-child chain
+      // get last one, the prior modules should already be published
+      IModule childModule = modules[modules.length - 1];
+      // XXX: Properties moduleUrls = server.loadModulePublishLocations();
+      Properties moduleUrls = new Properties(); // XXX: =
+      // server.loadModulePublishLocations();
+      // get as j2ee
+      IJ2EEModule childJ2EEModule = (IJ2EEModule) childModule.loadAdapter(IJ2EEModule.class, monitor);
+      if (childJ2EEModule != null && childJ2EEModule.isBinary()) {
+        publishArchiveModule(deployPath, moduleUrls, statusList, monitor, childModule);
+      } else {
+        publishDir(deployPath, moduleUrls, statusList, monitor, childModule);
+      }
+      // XXX: server.saveModulePublishLocations(moduleUrls);
+    }
+    checkStatuses(statusList);
+    server.setModulePublishState2(modules, IServer.PUBLISH_STATE_NONE);
+  }
+
+  private IStatus newErrorStatus(String message) {
```

Comment:
getClass() is non-static. Its a private method and dont think its worth it to hard code the name